### PR TITLE
Modularise `gutenbergOptInOut` state

### LINKED ISSUE #42438: Modularise `gutenbergOptInOut`
The `gutenbergOptInOut` portion of the state tree needs to be modularised. See the [modularised state documentation](https://github.com/Automattic/wp-calypso/blob/master/docs/modularized-state.md) for more details.

### PR DESCRIPTION
**Pull request by sgomes**
This PR is one of many working on modularising state in Calypso. This one handles `gutenbergOptInOut` state.

For more details on state modularisation, see the [modularised state documentation](https://github.com/Automattic/wp-calypso/blob/master/docs/modularized-state.md) and p4TIVU-9lM-p2.

Note: I added all the reviewers that GitHub suggested. Please feel free to ignore the review request or pull in someone else if you're not comfortable reviewing this PR. Thank you!

Fixes #42438.

#### Changes proposed in this Pull Request

* Modularise `gutenbergOptInOut` state.

#### Testing instructions

The lazy loading aspects of this PR aren't easy to test here, since this state seems to be loaded early on by Calypso, as part of other editor-related state.

As for the rest of the functionality, please smoke test Gutenberg opt-in/opt-out state and verify that it continues to work correctly. I couldn't find any issues, but my testing was very limited.

### CODE REVIEW

**Comment by sgomes** [comment]
Note: the linting errors are expected, since I'm modifying the root reducer, which uses linting errors to point out unmodularised state in an attempt to prevent more of it from being created.

**Comment by sgomes** [comment]
Rebased to fix conflict.

**Review by mmtr — COMMENTED** [review]
It worked as expected for me. Tested a bunch of scenarios:

- Switch to the classic editor from the "More" menu.
- Switch to the classic editor from the inline help.
- Switch to the block editor from the sidebar notice.
- Switch to the block editor from the modal dialog.
- Switch to the block editor from the deprecation card in My Home.

@Automattic/explorers or @Automattic/cylon may be more familiar with the current Gutenberg opt-in/opt-out flow though, so I'll let them to chime in.

**Comment by noahtallen** [comment]
Cylon hasn't touched this in a while, but it'd be good to get a +1 from @Automattic/explorers since they are making changes in this space for the editor deprecation project

**Comment by sgomes** [comment]
Thank you, @mmtr, @noahtallen!

@Automattic/explorers: any chance you can take a look?

**Comment by creativecoder** [comment]
Flagging for @glendaviesnz and @aaronrobertshaw , please give this a review today, if you can.

**Review by glendaviesnz — APPROVED** [review]
This all tested fine for me switching to and from block editor in the various places this is possible, and I also checked that it doesn't effect the upcoming editor deprecation work, and with this PR applied the deprecated users still do not see any options to switch to classic as expected.

**Comment by aaronrobertshaw** [comment]
This tested well for me as well. I could switch to and from the block editor via the:

- "More" menu
- Inline help popover
- Notices, dialogs and deprecation card

As a user within the editor deprecation group I still did not see any options to switch to classic which is as it should be.

**Comment by sgomes** [comment]
Thank you for handling the merge while I was AFK, @scinos! 🙇

**Pull request merged**
